### ISSUE TITLE
ZeroBin command-line interface

### ISSUE THREAD

**Issue by ziirish** [sebsauvage/ZeroBin#48]
Hi Seb,

I'm a little off topic but I didn't know how to reach you (I didn't search very long though).

I just wrote a cli for ZeroBin. It is a PoC/Draft and it probably should be enhanced but here it is: https://github.com/ziirish/zerocli

You can both paste or retrieve data. It is heavy as it requires rhino to execute javascript (as I couldn't find some openssl based implementation of the sjcl.js library) but it works.

**Comment by marsjaninzmarsa** [comment]
Why not Node.js?

**Comment by ziirish** [comment]
Hi,

I haven't used node.js because I don't know anything about it.
Now, if you use v8 instead of rhino, I *think* it's more or less equivalent to
using node (it uses the same engine) as explained
[here](https://en.wikipedia.org/wiki/Node.js):

> Node.js uses the Google V8 JavaScript engine to execute code

It means you could easily adapt this piece of code to use nodejs.


But the point was to be able to easily paste things from a headless machine
where you cannot always install whatever you want. It is the reason why I wrote
another version of zerocli that does not rely on any *huge* dependency
([zerocli2](https://github.com/ziirish/zerocli2)). For this to work, I also had
to fork the original ZeroBin code to use a cryptographic lib compatible with
openssl: [ZeroBin2](https://github.com/ziirish/ZeroBin2).

**Comment by marsjaninzmarsa** [comment]
Yes, Node is just better solution because much more people have it already installed and met dependency is cool dependency.

>It means you could easily adapt this piece of code to use nodejs.
I'll try. :smiley: 

>But the point was to be able to easily paste things from a headless machine where you cannot always install whatever you want. It is the reason why I wrote another version of zerocli that does not rely on any *huge* dependency

That's not the solution for me, because I have root on my own machine, I have even Node on my own machine and I wanna send things from my command line to **existing** servers.
BTW... why not PR the original ZeroBin with your code (and ofc switch in panel to select witch lib should be used for creating new pastes, and backward compatibility force using SJCL to decrypting existing pastes)?

**Comment by ziirish** [comment]
> That's not the solution for me, because I have root on my own machine, I have even Node on my own machine and I wanna send things from my command line to **existing** servers.

In your case you can do whatever you want with your server.
In my case I cannot do whatever I want with hundreds of servers.

I never said my solution was the best, it just fits my needs and I share it with
whomever can/want to use it, that's it.
That's also the reason why I maintain several distinct projects (zerocli,
zerocli2 and ZeroBin2) so that everybody is free to use either one or the other
solution.

> BTW... why not PR the original ZeroBin with your code (and ofc switch in panel to select witch lib should be used for creating new pastes, and backward compatibility force using SJCL to decrypting existing pastes)?

Have you seen the PR list? This project does not seem to be maintained anymore.
I have a PR waiting there for more than 2 years.

**Comment by marsjaninzmarsa** [comment]
>In your case you can do whatever you want with your server.

I don't have ZeroBin server (yet), so I can't. I just wanna do `cat something.txt | zerobin | xclippatch` and I'm looking for solution. :smile:

>I never said my solution was the best, it just fits my needs and I share it with whomever can/want to use it, that's it.

And it's great! Done is better than perfect. Your solution may be not the best, but it's working. :smiley:

>That's also the reason why I maintain several distinct projects (zerocli, zerocli2 and ZeroBin2) so that everybody is free to use either one or the other solution.

I'm just a fan of interoperability and not fan of fragmentation. I like your approach in ZeroBin2, but I wanna see it as configuration option or even selector for user, not separate project, because of last sentence.

>Have you seen the PR list? This project does not seem to be maintained anymore. I have a PR waiting there for more than 2 years.

Seen. Last accepted PR was sent 1.5y ago, merged same day, but it was security fix (#75).
Despite of that, I wanna try. If @sebsauvage don't answer, I wanna fork it. Or contribute to [elrido/ZeroBin](https://github.com/elrido/ZeroBin) (looks like in pretty active development, best regards @elrido!). Or to your ZeroBin2. Don't know yet. But in all cases, that about I will care most, will be interoperability and preventing of fragmentation. Abandoning projects is normal in open-source, unfortunately. But what is beautiful in open-source is the fact, that if project is abandoned and idea is god, **always** someone can get it and maintain it. In fact, that was nearly always more than one person, and what is challenging is to get this people together, but it is again nearly always worth it, because few people can do things better than one. And when project creator wakes up, they see project that can take over the legacy as official continuation or so...

**Comment by ziirish** [comment]
> I'm just a fan of interoperability and not fan of fragmentation. I like your approach in ZeroBin2, but I wanna see it as configuration option or even selector for user, not separate project, because of last sentence.
> 
> Seen. Last accepted PR was sent 1.5y ago, merged same day, but it was security fix (#75).
> Despite of that, I wanna try. If @sebsauvage don't answer, I wanna fork it. Or contribute to [elrido/ZeroBin](https://github.com/elrido/ZeroBin) (looks like in pretty active development, best regards @elrido!). Or to your ZeroBin2. Don't know yet. But in all cases, that about I will care most, will be interoperability and preventing of fragmentation. Abandoning projects is normal in open-source, unfortunately. But what is beautiful in open-source is the fact, that if project is abandoned and idea is god, **always** someone can get it and maintain it. In fact, that was nearly always more than one person, and what is challenging is to get this people together, but it is again nearly always worth it, because few people can do things better than one. And when project creator wakes up, they see project that can take over the legacy as official continuation or so...

I did not know about this fork when I made my changes. I'll try to backport them
and make a PR to this fork then.

**Comment by marsjaninzmarsa** [comment]
>I did not know about this fork when I made my changes.

Because it's new project, as I see.

>I'll try to backport them and make a PR to this fork then.

Great!
And I wanna work on some kind of universal REST API for ZeroBin and 0bin, like I wrote in #43. Both projects are modular and extendable, so I'm thinking about something like JSON schema with info about accepted encryption formats (SJCL/Crypto-js/etc), encrypter parameters and endpoints, so client can download it and compare with support provided by itself to choose the best one. :smiley:
